designs/0034-jacobian.md

```diff
@@ -0,0 +1,320 @@
+ upper_bound {
+   real constrain(real x, real upper_bound) {
+     jacobian += x;
+     return upper_bound - exp(x);
+   }
+   real constrain(vector x, real upper_bound) {
+     jacobian += sum(x);
+     return upper_bound - exp(x);
+   }
+   real unconstrain(real x, real upper_bound) {
+     return log(upper_bound - x);
+   }
+   // Used for data
+   int validate(real x, real upper_bound) {
+     // For most this will be
+     return x < upper_bound;
+   }
+ }
+}
+data {
+ real a;
+ real u_bound;
+ int N;
+}
+
+
+transformed data {
+ // Data will use the validate() function from the user-defined constraint
+ real<udc=upper_bound(u_bound)> a_validated;
+}
+
+
+parameters {
+ // User-defined constraint applied to parameters alpha and b
+ real<udc=upper_bound(u_bound)> alpha;
+ vector<udc=upper_bound(u_bound)> b;
+ // In these cases this is the same as calling
+ // real<upper=u_bound> alpha;
+ // vector<upper=u_bound>[N] b;
+}
+```
+
+In this example, upper_bound is a user-defined constraint that is applied to the data and parameters. The validate() function from upper_bound is used to validate a_validated in the transformed data block, and the transformations defined in upper_bound are applied to alpha and b in the parameters block.
+
+### Error Messages
+
+If the `constrain` function of a constraint does not increment `jacobian` a warning is thrown to inform users. In these cases it is better to use a standard function as a new constraint type is most likely not necessary.
+
+```md
+Warning: {CONSTRAINT_TYPE} does not increment the log of the absolute of the determinant of the jacobian.
+It is recommended to instead use a function in `transformed parameters` instead of a custom constraint.
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The first thing we would do is add the `jacobian` keyword to the Stan language. Like `target`, `jacobian` would only be available on the left hand side of statements to be accumulated into. Its domain is restricted to the `transformed parameters` and `constraints` block.
+
+This section will break down the model given in the summary section to show what the Stan compiler requires piece by piece. First, let's look at the `constraints` block where we define our constrain and unconstrain functions
+
+## Constraint Block
+
+```stan
+constraints {
+ upper_bound {
+   constrain(real x, real upper_bound) {
+     jacobian += x;
+     return upper_bound - exp(x);
+   }
+   real unconstrain(real x, real upper_bound) {
+     return log(upper_bound - x);
+   }
+   // Used for data
+   int validate(real x, real upper_bound) {
+     // For most this will be
+     return x < upper_bound;
+   }
+ }
+}
+```
+
+The new `constraints` block will need to be added to the compiler's parser, validator, mir, and c++ mir. The parser will transpile each of the constraints to c++ functions which start with the name of the constraint. For the example above the signatures would be
```

Comment:
Always list out acronyms in full before using them the first time.  So that'd be "middle implementation layer (MIR)".  And unless it's "cdf" or "pdf", capitalize acronyms.

Reply:
👍